controller/src/controller.rs

```diff
@@ -0,0 +1,149 @@
+use super::{
+    error::{self, Result},
+    statemachine::determine_next_node_spec,
+};
+use models::node::{BottlerocketNode, BottlerocketNodeClient, BottlerocketNodeState};
+
+use kube::runtime::reflector::Store;
+use kube::ResourceExt;
+use snafu::ResultExt;
+use std::collections::BTreeMap;
+use tokio::time::{sleep, Duration};
+use tracing::{event, instrument, Level};
+
+// Defines the length time after which the controller will take actions.
+const ACTION_INTERVAL: Duration = Duration::from_secs(2);
+
+/// The BrupopController orchestrates updates across a cluster of Bottlerocket nodes.
+pub struct BrupopController<T: BottlerocketNodeClient> {
+    node_client: T,
+    brn_reader: Store<BottlerocketNode>,
+}
+
+impl<T: BottlerocketNodeClient> BrupopController<T> {
+    pub fn new(node_client: T, brn_reader: Store<BottlerocketNode>) -> Self {
+        BrupopController {
+            node_client,
+            brn_reader,
+        }
+    }
+
+    /// Returns a list of all `BottlerocketNode` objects in the cluster.
+    fn all_nodes(&self) -> Vec<BottlerocketNode> {
```

Comment:
Curious if we need this function here? Since you mentioned in my PR before that those kind functions essentially just getter functions written as top-level functions. Should we just use self.brn_reader.state() when we need it? : )

Reply:
Good question! In this case, this function *is* a getter function. Getter functions are fine, but they should probably be provided as a method on the struct that they are retrieving data from. When I mentioned a top-level function, I meant that the function in the other PR was implemented on the module, rather than the struct it was accessing.

To clarify, I think either making an inner value `pub` could be appropriate if it's okay for customers to reach in and change the value as well as retrieve it, or something like this:
```
struct A {
    inner_value: String
}

impl A {
    fn inner_value(&self) -> &str {
        &self.inner_value
    }
}
```

Is more appropriate than something like this:

```
struct A {
    inner_value: String,
}

pub fn get_inner_value(my_a: &A) -> &str {
    a.inner_value
}
```